Add websockets and data messages (#42)
fixie-ai/ultradox: docs/src/content/docs/datamessages.mdx

---

Data messages are used to communicate non-audio information between your client and an Ultravox server during calls. These messages work across WebRTC data channels and WebSocket connections.

All messages are JSON objects with camelCase keys containing:
- A required `type` field identifying the message type
- Additional fields specific to each message type

## Messages at a Glance
This table provides all messages at a glance. Details on each message type appears below. Sender indicates client or server message. Client messages are sent from the client to the server. Server messages are sent from the server to the client.
| Message | Sender | Description |
| --------------------------------------------- | ------ | ---------------------------------------------------- |
| [Ping](#ping)                                 | Client | Measures round-trip data latency.                    |
| [Pong](#pong)                                 | Server | Server reply to a ping message.                      |
| [State](#state)                               | Server | Indicates the server's current state.                |

> | [Transcript](#transcript)                     | Server | Contains text transcripts of audio.                  |

Reviewer: ```suggestion
| [Transcript](#transcript)                     | Server | Contains text for an utterance made during the call.                  |
```

| [InputTextMessage](#inputtextmessage)         | Client | Used to send a user message to the agent via text.   |
| [SetOutputMedium](#setoutputmedium)           | Client | Sets server's output medium to text or voice.        |
| [ClientToolInvocation](#clienttoolinvocation) | Server | Asks the client to invoke a client tool.             |
| [ClientToolResult](#clienttoolresult)         | Client | Contains the result of a client tool invocation.     |
| [Debug](#debug)                               | Server | Useful for application debugging.                    |
| [PlaybackClearBuffer](#playbackclearbuffer)   | Server | Used to clear buffered output audio. WebSocket only. |

## Ping
A message sent by the client to measure round-trip data message latency.
- `type: "ping"` 
- `timestamp`: Float. Client timestamp for latency measurement.

## Pong
A message sent by the server in response to a PingMessage. The timestamp is copied from the PingMessage.
- `type: "pong"`
- `timestamp`: Float. Matching ping timestamp.

## State
A message sent by the server to indicate its current state.
- `type: "state"`
- `state`: Current session state

## Transcript

> A message containing text transcripts of user and agent audio utterances.

Reviewer: ```suggestion
A message containing text transcripts of user and agent utterances.
```

- `type: "transcript"`
- `role`: "user" or "agent". Who emitted the utterance.
- `medium`: "text" or "voice". The medium through which the utterance was emitted.
- `text`: String. Full transcript text (exclusive with delta). The full text of the transcript so far. Either this or delta will be set.
- `delta`: String. Incremental transcript update (exclusive with text). The additional transcript text added since the last agent transcript message.

> - `final`: Boolean. Whether more updates expected. Whether to expect additional transcript messages for this conversation round.

Reviewer: ```suggestion
- `final`: Boolean. Whether more updates are expected for this utterance.
- `ordinal`: int. Used for ordering transcripts within a call.
```

## InputTextMessage
A user message sent via text.
- `type: "input_text_message"`
- `text`: String. The content of the user message.

## SetOutputMedium
Message sent by the client to set the server's output medium.
- `type: "set_output_medium"`
- `medium`: Either "voice" or "text".

## ClientToolInvocation
Sent by the server to ask the client to invoke a client-implemented tool with the given parameters. The client is expected to send back a ClientToolResultMessage with a matching invocation_id.
- `type: "client_tool_invocation"`
- `tool_name`: String. Tool to invoke
- `invocation_id`: String. Unique invocation ID
- `parameters`: Dict[String, Any]. Tool parameters

## ClientToolResult
Contains the result of a client-implemented tool invocation.
- `type: "client_tool_result"`
- `invocation_id`: String. Matches corresponding invocation.

> - `result`: String. Tool execution result (optional).

Reviewer: ```suggestion
- `result`: String. Tool execution result. Often a JSON string. May be omitted for errors.
```

- `response_type`: String. Defaults to "tool-response".

> - `error_type`: "undefined" or "implementation-error" (optional).

Reviewer: ```suggestion
- `error_type`: Optional string. Should be omitted when result is set. Otherwise, should be "undefined" if the a tool with the given name does not exist or "implementation-error" otherwise.
```

- `error_message`: String. Error details if failed (optional).

## Debug
A message sent by the server to communicate debug information.
- `type: "debug"`
- `message`: String. Debug information
- Disabled by default

## PlaybackClearBuffer
Message sent by our server to clear buffered output audio. Integrators should drop as much unplayed output audio as possible in order for interruptions to function properly.
- `type: "playback_clear_buffer"`
- WebSocket connections only

> - Signals client to clear unplayed audio

Reviewer: ```suggestion
```

---
title: "Data Messages"
description: Protocol documentation for messages exchanged between client and server during Ultravox calls.
---

Data messages are used to communicate state, transcripts, and control information between the client and server during Ultravox calls. These messages work across WebRTC data channels and WebSocket connections.

## Message Structure

All messages are JSON objects containing:
- A required `type` field identifying the message type
- Additional fields specific to each message type
- Messages follow camelCase naming convention

> ## Client Messages

Reviewer: Related to my comment above, I think it's more useful to order these by use case than any other categorization.  For example, if I want to know about `Ping` messages, the next message type I'll want to know about is almost certainly `Pong` messages.

Messages sent from client to server:

### Input Messages

> - **InputTextMessage**

Reviewer: Might be useful to have a brief description of what each message is before going into its schema.  E.g.:

- **InputTextMessage** - A user message sent via text.
    + `type: "input_text_message"`
    + `text`: The content of the user message
    
I think I included docstrings for all the messages in data_messages.py that ought to give a good starting point.

Author: done.

  - `type: "input_text_message"`
  - `text`: Text content from user

### Control Messages
- **Ping**
  - `type: "ping"` 
  - `timestamp`: Client timestamp for latency measurement

- **SetOutputMedium**
  - `type: "set_output_medium"`
  - `medium`: Either "voice" or "text"

### Tool Messages
- **ClientToolResult**
  - `type: "client_tool_result"`
  - `invocation_id`: Matches corresponding invocation
  - `result`: Tool execution result (optional)
  - `response_type`: Defaults to "tool-response"
  - `error_type`: "undefined" or "implementation-error" (optional)
  - `error_message`: Error details if failed (optional)

## Server Messages

Messages sent from server to client:

### Transcript Messages

> - **Transcript**

Reviewer: This will also have `ordinal` before we start using this message (at least with the SDKs).  @petersalas - Can you link to your PR when it's ready so we can make sure the docs reflect the related changes please?

Author: did Peter's PR land? holding on this for now.

> - **Transcript**

Reviewer: Related to my comment about some overview information for each message, it will eventually be useful to mention the API version in which a message was introduced.  (Probably not worth documenting version 0, so we don't need this yet but worth keeping in mind.)

Author: ok sounds good

  - `type: "transcript"`
  - `role`: "user" or "agent"
  - `medium`: "text" or "voice" 
  - `text`: Full transcript text (exclusive with delta)
  - `delta`: Incremental transcript update (exclusive with text)
  - `final`: Whether more updates expected

### State & Control
- **State**
  - `type: "state"`
  - `state`: Current session state

- **Pong**
  - `type: "pong"`
  - `timestamp`: Matching ping timestamp

### Tool Messages
- **ClientToolInvocation**
  - `type: "client_tool_invocation"`
  - `tool_name`: Tool to invoke
  - `invocation_id`: Unique invocation ID
  - `parameters`: Tool parameters

### WebSocket-Specific
- **PlaybackClearBuffer**
  - `type: "playback_clear_buffer"`
  - WebSocket connections only
  - Signals client to clear unplayed audio

### Debug & Metrics
- **Debug**
  - `type: "debug"`
  - `message`: Debug information
  - Disabled by default

> - **Latency**

Reviewer: I don't think we want to document this one.  It exists for our own internal debugging, but I'm not confident enough in the accuracy to be comfortable with anyone else trying to use it.  (It's also very tied to the component stack we're actively moving away from.)

Author: removed

  - `type: "latency"`
  - `kind`: Metric type
  - `value`: Measurement in milliseconds

> :::note[Message Size Limit]

Reviewer: I suspect this is only relevant for client tool responses.  Maybe move this to the tools page?  (This also isn't accurate otherwise.  The issue appears to be that LiveKit transparently drops large messages even though we send them "reliably". That happens to cause a timeout for client implemented tool calls because our server requires a response within a reasonable time frame and the response getting dropped by LiveKit means we'll always hit that time limit.)

Author: ok moved to the SDK page where we talk about client tools
